BinaryTree.java

```diff
@@ -15,99 +24,133 @@ public BinaryTree(BinaryTreeNode<T> root) {
+            System.err.println("This value already exists");
             return;
         }
 
         // Check if the parent exists in the tree
-        if (!this.isExist(parent)) {
-            System.out.println("This parent isn't exist in the tree");
+        if (this.exists(parent) == null) {
+            System.err.println("This parent isn't exist in the tree");
             return;
         }
-
-        insert(value, this.findNode(parent), toLeft);
+        insert(value, this.exists(parent));
     }
 
     // The function create a left (left or right) to the parent node
-    public void insert(T value, BinaryTreeNode<T> parent, boolean toLeft) {
-
-        if (toLeft) {
+    public void insert(T value, BinaryTreeNode<T> parent) {
+        if (m_insertChild == InsertChild.Left) {
             if (parent.getLeftChild() == null) {
                 parent.setLeftChild(value);
             } else {
-                System.out.println("There is already left child");
+                System.err.println("There is already left child");
                 return;
             }
         } else {
             if (parent.getRightChild() == null) {
                 parent.setRightChild(value);
             } else {
-                System.out.println("There is already right child");
+                System.err.println("There is already right child");
                 return;
             }
         }
     }
-}
+
+    // The function receives a Binary Tree and a key
+    // Prints all the ancestors of the key from the root
+    public List<T> printAncestors(BinaryTree<T> tree, T key) {
+        if (tree == null) {
+            System.err.println("The tree is empty");
+            return null;
+        }
+        BinaryTreeNode root = tree.getRoot();
+
+        if (tree.exists(key) == null) {
+            System.err.println("There is no such key in the tree");
+            return null;
+        }
+        List<T> ancestors = new ArrayList<>();
+        ancestors = printAncestors(ancestors, root, key);
+
+        for(T ancestor : ancestors) {
+            System.out.print(ancestor + " ");
+        }
+        return ancestors;
+    }
+
+    public List<T> printAncestors(List<T> ancestors, BinaryTreeNode<T> root, T key) {
+        if (root.getData() == key) {
+            return ancestors;
+        }
+
+        if (exists(root, key) != null) {
+            ancestors.add(0, (T) root.getData());
+
+            if (exists(root.getRightChild(), key) != null) {
+                printAncestors(ancestors, root.getRightChild(), key);
+            }
+
+            if (exists(root.getLeftChild(), key) != null) {
+                printAncestors(ancestors, root.getLeftChild(), key);
+            }
+        }
+        return ancestors;
+    }
+
+    // The function receives a Binary Tree and two nodes
+    // Return the lowest common ancestor
+    public BinaryTreeNode<T> lowestCommonAncestor(BinaryTree<T> tree, BinaryTreeNode<T> firstNode,

```

Comment:
I might have misunderstood the question. Does the function need to get two keys and return the key of the lower common ancestor?

How would you like to be informed about failure? - As a user or a developer?